packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -238,43 +285,142 @@ contract FundingRateStore is FundingRateStoreInterface, Testable, Lockable {
     }
 
+    /**
+     * @notice Publishes any expired pending proposals and pays out proposal rewards if neccessary.
+     * @param perpetual Contract to check proposed funding rates for.
+     */
+    function withdrawProposalRewards(address perpetual) external nonReentrant() publishAndWithdrawProposal(perpetual) {}
+
+    /**
+     * @notice Set the reward rate (per second) for a specific `perpetual` contract.
+     * @dev Callable only by the Perpetual contract.
+     */
+    function setRewardRate(address perpetual, FixedPoint.Unsigned memory rewardRate)
+        external
+        override
+        nonReentrant()
+        publishAndWithdrawProposal(perpetual)
+    {
+        require(msg.sender == perpetual, "Caller not perpetual");
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(perpetual);
+        fundingRateRecord.rewardRatePerSecond = rewardRate;
+
+        // Set last propose time to current time since rewards will start accruing from here on out.
+        fundingRateRecord.proposeTime = getCurrentTime();
+
+        emit ChangedRewardRate(perpetual, rewardRate.rawValue);
+    }
+
+    /**
+     * @notice Helpful method for proposers who want to calculate their potential rewards and useful for testing.
+     */
+    function calculateProposalRewardPct(
+        address perpetual,
+        uint256 startTime,
+        uint256 endTime,
+        FixedPoint.Signed memory proposedRate,
+        FixedPoint.Signed memory currentRate
+    ) external view returns (FixedPoint.Unsigned memory reward) {
+        return _calculateProposalRewardPct(perpetual, startTime, endTime, proposedRate, currentRate);
+    }
+
     /****************************************
      *         INTERNAL FUNCTIONS           *
      ****************************************/
 
+    // TODO: Do we need to add a reentrancy guard to this method because it makes external calls?
+    function _publishRateAndWithdrawRewards(address perpetual) internal {
+        FundingRateRecord storage fundingRateRecord = _getFundingRateRecord(perpetual);
+
+        // Check if proposal liveness has expired
+        if (_getProposalState(fundingRateRecord.proposal) != ProposalState.Expired) {
+            return;
+        }
+
+        // Publish rate and proposal time.
+        fundingRateRecord.rate = fundingRateRecord.proposal.rate;
+        fundingRateRecord.proposeTime = fundingRateRecord.proposal.time;
+
+        // Calculate reward for proposer using saved reward rate and current perpetual PfC.
+        FixedPoint.Unsigned memory pfc = AdministrateeInterface(perpetual).pfc();
+        FixedPoint.Unsigned memory rewardRate = fundingRateRecord.proposal.rewardRate;
+        FixedPoint.Unsigned memory reward = rewardRate.mul(pfc);
+
+        // Pull reward payment from Perpetual, and transfer (payment + final fee bond) to proposer.
+        // TODO: Should we add a try-catch around these external calls incase the `perpetual.withdrawFundingRateFees`
```

Comment:
What would the catch state do? It would make sense to do this perhaps as this function is applied as a modifier in a few places and so it could cause other logic to revert & that might be a broken state. For example, if this function is reverting due to the perpetual being unable to withdraw funds for whatever reason it could be possible that the funding rate store's pending funding rate can not be disputed.

Reply:
Yeah I think the catch state should maybe just emit an event, but it should make sure that the Perpetual.withdrawFees doesn't throw for any reason

Reply:
I added a catch statement + tests to make sure a reverting `withdrawFundingRateFees` doesn't freeze the contract